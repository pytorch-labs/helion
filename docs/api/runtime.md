# Runtime Module

The `helion.runtime` module handles kernel execution and configuration management.

```{eval-rst}
.. currentmodule:: helion.runtime
```

## Key Classes

The runtime module provides the core execution infrastructure. For detailed documentation of individual classes, see:

- {doc}`config` - Configuration management
- {doc}`kernel` - Kernel execution
- {doc}`settings` - Compilation settings

## Utility Functions

```{eval-rst}
.. currentmodule:: helion.runtime

.. autofunction:: triton_wait_signal

.. autofunction:: get_num_sm

.. autofunction:: set_triton_allocator
```
