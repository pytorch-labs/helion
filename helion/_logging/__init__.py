from __future__ import annotations

from ._internal import LazyString as LazyString
from ._internal import init_logs as init_logs
