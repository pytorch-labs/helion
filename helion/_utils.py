from __future__ import annotations

import collections

counters: collections.defaultdict[str, collections.Counter[str]] = (
    collections.defaultdict(collections.Counter)
)
