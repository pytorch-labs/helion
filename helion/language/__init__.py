from __future__ import annotations

from .constexpr import ConstExpr as constexpr  # noqa: F401
from .creation_ops import full as full
from .creation_ops import zeros as zeros
from .loops import tile as tile
from .memory_ops import load as load
from .memory_ops import store as store
from .view_ops import subscript as subscript
