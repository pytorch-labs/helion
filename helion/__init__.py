from __future__ import annotations

from . import _logging
from . import exc
from . import language
from . import runtime
from .runtime import Config
from .runtime import Kernel
from .runtime import kernel
from .runtime import kernel as jit  # alias
from helion.runtime.settings import Settings
from helion.runtime.settings import set_default_settings

__all__ = [
    "Config",
    "Kernel",
    "Settings",
    "exc",
    "jit",
    "kernel",
    "language",
    "runtime",
    "set_default_settings",
]

_logging.init_logs()
