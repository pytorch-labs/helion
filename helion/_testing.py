from __future__ import annotations

import importlib
import sys
from typing import TYPE_CHECKING

import torch

from .runtime.config import Config

if TYPE_CHECKING:
    from pathlib import Path
    import types

    from .runtime.kernel import Kernel


DEVICE = torch.device("cuda")


def import_path(filename: Path) -> types.ModuleType:
    module_name = f"{__name__}.{filename.stem}"
    if module_name not in sys.modules:
        spec = importlib.util.spec_from_file_location(module_name, filename)
        assert spec is not None
        module = importlib.util.module_from_spec(spec)
        assert spec.loader is not None
        spec.loader.exec_module(module)
        sys.modules[module_name] = module
    return sys.modules[module_name]


def code_and_output(
    # pyre-ignore[11]
    fn: Kernel,
    args: tuple[object, ...],
    **kwargs: object,
) -> tuple[str, object]:
    if kwargs:
        config = Config(**kwargs)
    elif fn.configs:
        (config,) = fn.configs
    else:
        config = fn.bind(args).config_spec.default_config()
    code = fn.bind(args).to_triton_code(config)
    compiled_kernel = fn.bind(args).compile_config(config)
    try:
        result = compiled_kernel(*args)
    except Exception:
        sys.stderr.write(f"Failed to run kernel:\n{code}\n")
        raise
    return code, result
