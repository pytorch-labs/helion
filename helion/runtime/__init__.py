from __future__ import annotations

import functools

import torch
import triton

from .config import Config as Config
from .kernel import Kernel as Kernel
from .kernel import autotune as autotune
from .kernel import kernel as kernel
from .triton_helpers import triton_send_signal as triton_send_signal
from .triton_helpers import triton_wait_multiple_signal as triton_wait_multiple_signal
from .triton_helpers import triton_wait_signal as triton_wait_signal


def _alloc_fn(size: int, alignment: int, stream: int | None) -> torch.Tensor:
    return torch.empty(size, device="cuda", dtype=torch.int8)


@functools.cache
def set_triton_allocator() -> None:
    try:
        from triton.runtime._allocation import NullAllocator
        from triton.runtime._allocation import _allocator

        if not isinstance(_allocator, NullAllocator):
            return
    except ImportError:
        pass
    triton.set_allocator(_alloc_fn)


def get_num_sm(device: torch.device) -> int:
    """
    Get the number of streaming multiprocessors (SMs) for the specified device.

    Args:
        device: Device to query.

    Returns:
        Grid size to use for a persistent kernel on the device.
    """
    assert device.type == "cuda", "TODO: implement for other devices"
    return torch.cuda.get_device_properties(device.index).multi_processor_count


def default_launcher(
    triton_kernel: triton.JITFunction,
    grid: tuple[int, ...],
    *args: object,
    num_warps: int,
    num_stages: int,
) -> object:
    """Default launcher function that executes the kernel immediately."""
    return triton_kernel.run(
        *args, grid=grid, warmup=False, num_warps=num_warps, num_stages=num_stages
    )
