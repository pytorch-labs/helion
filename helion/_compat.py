from __future__ import annotations

import functools
import importlib

import torch
from torch._inductor.runtime.hints import DeviceProperties
from torch._inductor.utils import triton_type
from triton.backends.compiler import GPUTarget
import triton.language as tl


def supports_tensor_descriptor() -> bool:
    # call private func we can patch in testing
    return _supports_tensor_descriptor()


@functools.cache
def _supports_tensor_descriptor() -> bool:
    if not torch.cuda.is_available():
        return False
    major, _ = torch.cuda.get_device_capability(torch.cuda.current_device())
    if major < 9:
        return False
    try:
        return get_triton_tensor_descriptor_import_path() is not None
    except ImportError:
        return False


@functools.cache
def get_triton_tensor_descriptor_import_path() -> str:
    """Attempt to import TensorDescriptor object from known Triton modules."""
    possible_modules = [
        "triton.tools.experimental_descriptor",
        "triton.tools.tensor_descriptor",
    ]
    for module_name in possible_modules:
        try:
            module = importlib.import_module(module_name)
            if hasattr(module, "TensorDescriptor"):
                return f"from {module_name} import TensorDescriptor"
        except ImportError:
            continue
    raise ImportError("TensorDescriptor not found in any of the known Triton modules.")


@functools.cache
def torch_dtype_to_tl(torch_dtype: torch.dtype) -> object:
    """Return the `triton.language` dtype that matches a `torch.dtype`."""
    name_str = triton_type(torch_dtype).replace("tl.", "")
    return getattr(tl, name_str)


@functools.cache
def min_dot_size(
    device: torch.device, lhs: torch.dtype, rhs: torch.torch.dtype
) -> tuple[int, int, int]:
    if device.type != "cuda":
        # TODO(jansel): support non-cuda properly
        return (16, 16, 16)

    from triton.backends.nvidia.compiler import min_dot_size as min_dot_size_cuda

    props = DeviceProperties.create(device)
    return min_dot_size_cuda(
        GPUTarget(
            backend=props.type,
            arch=props.cc,
            warp_size=props.warp_size or 32,
        )
    )(torch_dtype_to_tl(lhs), torch_dtype_to_tl(rhs))


def warps_to_threads(num_warps: int) -> int:
    if torch.cuda.is_available():
        props = DeviceProperties.create(
            torch.device("cuda", torch.cuda.current_device())
        )
        return num_warps * (props.warp_size or 32)
    return num_warps * 32
