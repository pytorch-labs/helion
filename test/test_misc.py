from __future__ import annotations

import unittest

from expecttest import TestCase
import pytest
import torch

import helion
from helion._testing import DEVICE
from helion._testing import code_and_output
import helion.language as hl


class TestMisc(TestCase):
    def test_decorator(self):
        def mydec(func):
            return func

        @mydec
        @helion.kernel
        def add1(x, y):
            x, y = torch.broadcast_tensors(x, y)
            out = torch.empty_like(x)
            for tile in hl.tile(out.size()):
                out[tile] = x[tile] + y[tile]
            return out

        @helion.kernel
        @mydec
        def add2(x, y):
            x, y = torch.broadcast_tensors(x, y)
            out = torch.empty_like(x)
            for tile in hl.tile(out.size()):
                out[tile] = x[tile] + y[tile]
            return out

        @mydec
        @helion.kernel(config=helion.Config(block_size=[4]))
        def add3(x, y):
            x, y = torch.broadcast_tensors(x, y)
            out = torch.empty_like(x)
            for tile in hl.tile(out.size()):
                out[tile] = x[tile] + y[tile]
            return out

        x = torch.randn(4, device=DEVICE)

        code_and_output(add1, (x, x))

        with pytest.raises(
            expected_exception=helion.exc.DecoratorAfterHelionKernelDecorator,
            match="Decorators after helion kernel decorator are not allowed",
        ):
            code_and_output(add2, (x, x))

        code_and_output(add3, (x, x))

    def test_inputs(self):
        @helion.kernel
        def kernel(a_list, b_dict, b_tuple):
            a0, a1 = a_list
            b0 = b_dict["b0"]
            (b1,) = b_tuple
            c0, c1 = torch.empty_like(a0), torch.empty_like(a1)
            for tile in hl.tile(a0.size()):
                c0[tile] = a0[tile] + b0[tile]
                c1[tile] = a1[tile] + b1[tile]
            return [c0, c1]

        x = torch.randn(4, device=DEVICE)
        code, result = code_and_output(kernel, ([x, x], {"b0": x}, (x,)))
        torch.testing.assert_close(result[0], 2 * x)
        torch.testing.assert_close(result[1], 2 * x)
        self.assertExpectedInline(
            code,
            """\
from __future__ import annotations

import torch
import triton
import triton.language as tl

@triton.jit
def _kernel_kernel(a0, c0, c1, a0_size_0, a0_stride_0, c0_stride_0, c1_stride_0, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = offset_0 + tl.arange(0, _BLOCK_SIZE_0).to(tl.int32)
    mask_0 = indices_0 < a0_size_0
    load = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    load_1 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_0 = load + load_1
    tl.store(c0 + indices_0 * c0_stride_0, v_0, mask_0)
    load_2 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    load_3 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_1 = load_2 + load_3
    tl.store(c1 + indices_0 * c1_stride_0, v_1, mask_0)

def kernel(a_list, b_dict, b_tuple):
    a0, a1 = a_list
    b0 = b_dict['b0']
    b1, = b_tuple
    c0, c1 = (torch.empty_like(a0), torch.empty_like(a1))
    _BLOCK_SIZE_0 = 4
    _kernel_kernel[triton.cdiv(a0.size(0), _BLOCK_SIZE_0),](a0, c0, c1, a0.size(0), a0.stride(0), c0.stride(0), c1.stride(0), _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return [c0, c1]

def _kernel_make_precompiler(a_list, b_dict, b_tuple):
    a0, a1 = a_list
    b0 = b_dict['b0']
    b1, = b_tuple
    c0, c1 = (torch.empty_like(a0), torch.empty_like(a1))
    _BLOCK_SIZE_0 = 4
    from helion.runtime.precompile_shim import make_precompiler
    return make_precompiler(_kernel_kernel)(a0, c0, c1, a0.size(0), a0.stride(0), c0.stride(0), c1.stride(0), _BLOCK_SIZE_0, num_warps=4, num_stages=3)""",
        )


if __name__ == "__main__":
    unittest.main()
