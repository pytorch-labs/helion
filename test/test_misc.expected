This file is automatically generated by assertExpectedJournal calls in test_misc.py.
Update expected outputs by running tests with the EXPECTTEST_ACCEPT=1 environment variable set.

--- assertExpectedJournal(TestMisc.test_inputs)
from __future__ import annotations

import torch
import triton
import triton.language as tl

@triton.jit
def _kernel_kernel(a0, o0, o1, a0_size_0, a0_stride_0, o0_stride_0, o1_stride_0, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < a0_size_0
    load = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    load_1 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_0 = load + load_1
    load_2 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_1 = v_0 + load_2
    load_3 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_2 = v_1 + load_3
    tl.store(o0 + indices_0 * o0_stride_0, v_2, mask_0)
    load_4 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    load_5 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_3 = load_4 + load_5
    load_6 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_4 = v_3 + load_6
    load_7 = tl.load(a0 + indices_0 * a0_stride_0, mask_0, other=0)
    v_5 = v_4 + load_7
    tl.store(o1 + indices_0 * o1_stride_0, v_5, mask_0)

def kernel(a_list, b_dict, b_tuple, c_named_tuple, d_dataclass):
    a0, a1 = a_list
    b0 = b_dict['b0']
    b1, = b_tuple
    c0, c1 = (c_named_tuple.x, c_named_tuple.y)
    d0, d1 = (d_dataclass.x, d_dataclass.y)
    o0, o1 = (torch.empty_like(a0), torch.empty_like(a1))
    _BLOCK_SIZE_0 = 4
    _kernel_kernel[triton.cdiv(a0.size(0), _BLOCK_SIZE_0),](a0, o0, o1, a0.size(0), a0.stride(0), o0.stride(0), o1.stride(0), _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return [o0, o1]

def _kernel_make_precompiler(a_list, b_dict, b_tuple, c_named_tuple, d_dataclass):
    a0, a1 = a_list
    b0 = b_dict['b0']
    b1, = b_tuple
    c0, c1 = (c_named_tuple.x, c_named_tuple.y)
    d0, d1 = (d_dataclass.x, d_dataclass.y)
    o0, o1 = (torch.empty_like(a0), torch.empty_like(a1))
    _BLOCK_SIZE_0 = 4
    from helion.runtime.precompile_shim import make_precompiler
    return make_precompiler(_kernel_kernel)(a0, o0, o1, a0.size(0), a0.stride(0), o0.stride(0), o1.stride(0), _BLOCK_SIZE_0, num_warps=4, num_stages=3)

--- assertExpectedJournal(TestMisc.test_tile_block_size_constexpr_fix)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from torch._inductor.runtime.triton_compat import libdevice

@triton.jit
def _test_tile_block_size_usage_kernel(out, x_size_0, out_stride_0, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < x_size_0
    _BLOCK_SIZE_0_ = _BLOCK_SIZE_0
    v_0 = _BLOCK_SIZE_0_.to(tl.int32)
    v_1 = indices_0 % v_0
    v_2 = tl.full([], 0, tl.int32)
    v_3 = v_1 != v_2
    v_4 = libdevice.signbit(v_1) != 0 if v_1.dtype is tl.float32 else v_1 < 0
    v_5 = libdevice.signbit(v_0) != 0 if v_0.dtype is tl.float32 else v_0 < 0
    v_6 = v_4 != v_5
    v_7 = v_3 & v_6
    v_8 = v_1 + v_0
    v_9 = tl.where(v_7, v_8, v_1)
    sub = -1 + _BLOCK_SIZE_0
    v_10 = sub.to(tl.int32)
    v_11 = v_9 == v_10
    v_12 = tl.full([], 0, tl.int64)
    v_13 = tl.full([], 1, tl.int64)
    v_14 = v_13[None]
    v_15 = v_12[None]
    v_16 = tl.where(v_11, v_14, v_15)
    v_17 = v_16.to(tl.int32)
    tl.store(out + indices_0 * out_stride_0, v_17, mask_0)

def test_tile_block_size_usage(x: torch.Tensor):
    out = torch.zeros_like(x, dtype=torch.int32)
    _BLOCK_SIZE_0 = 32
    _test_tile_block_size_usage_kernel[triton.cdiv(x.size(0), _BLOCK_SIZE_0),](out, x.size(0), out.stride(0), _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return out

def _test_tile_block_size_usage_make_precompiler(x: torch.Tensor):
    out = torch.zeros_like(x, dtype=torch.int32)
    _BLOCK_SIZE_0 = 32
    from helion.runtime.precompile_shim import make_precompiler
    return make_precompiler(_test_tile_block_size_usage_kernel)(out, x.size(0), out.stride(0), _BLOCK_SIZE_0, num_warps=4, num_stages=3)

--- assertExpectedJournal(TestMisc.test_torch_alloc)
from __future__ import annotations

import torch
import triton
import triton.language as tl

@triton.jit
def _fn_kernel(x, out, out_stride_0, x_stride_0, x_stride_1, m, n, _BLOCK_SIZE_1: tl.constexpr, _BLOCK_SIZE_0: tl.constexpr):
    pid_0 = tl.program_id(0)
    offset_1 = pid_0 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < m
    acc = tl.full([_BLOCK_SIZE_1, _BLOCK_SIZE_0], 0, tl.float32)
    for offset_0 in tl.range(0, n.to(tl.int32), step=_BLOCK_SIZE_0):
        indices_0 = offset_0 + tl.arange(0, _BLOCK_SIZE_0).to(tl.int32)
        mask_0 = indices_0 < n
        acc_copy = acc
        acc_copy_0 = acc_copy
        load = tl.load(x + (indices_1[:, None] * x_stride_0 + indices_0[None, :] * x_stride_1), mask_1[:, None] & mask_0[None, :], other=0)
        acc = acc_copy_0 + load
    sum_1 = tl.sum(acc, 1)
    tl.store(out + indices_1 * out_stride_0, sum_1, mask_1)

def fn(x: torch.Tensor):
    m, n = x.size()
    out = x.new_empty([m])
    _BLOCK_SIZE_1 = 64
    _BLOCK_SIZE_0 = 64
    _fn_kernel[triton.cdiv(m, _BLOCK_SIZE_1),](x, out, out.stride(0), x.stride(0), x.stride(1), m, n, _BLOCK_SIZE_1, _BLOCK_SIZE_0, num_warps=4, num_stages=3)
    return out

def _fn_make_precompiler(x: torch.Tensor):
    m, n = x.size()
    out = x.new_empty([m])
    _BLOCK_SIZE_1 = 64
    _BLOCK_SIZE_0 = 64
    from helion.runtime.precompile_shim import make_precompiler
    return make_precompiler(_fn_kernel)(x, out, out.stride(0), x.stride(0), x.stride(1), m, n, _BLOCK_SIZE_1, _BLOCK_SIZE_0, num_warps=4, num_stages=3)
